synapse/handlers/sso.py

```diff
@@ -701,8 +707,69 @@ async def _register_mapped_user(
         await self._store.record_user_external_id(
             auth_provider_id, remote_user_id, registered_user_id
         )
+
+        # Set avatar, if available
+        if attributes.picture:
+            await self.set_avatar(registered_user_id, attributes.picture)
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        try:
+            uid = UserID.from_string(user_id)
+
+            # HEAD request to find image size & mime type before download
+            response = await self._http_client.request("HEAD", picture_https_url)
+            if response.code != 200:
+                raise Exception("error sending HEAD request to get image size")
+
+            content_length = int(
+                response.headers.getRawHeaders(b"Content-Length")[0].decode("utf-8")
+            )
+            content_type = response.headers.getRawHeaders(b"Content-Type")[0].decode(
+                "utf-8"
+            )
+
+            # ensure picture size respects max_avatar_size defined in config
+            if self._profile_handler.max_avatar_size is not None:
+                if content_length > self._profile_handler.max_avatar_size:
+                    raise Exception("sso avatar too big in size")
+
+            # ensure picture is of allowed mime type
+            if (
+                self._profile_handler.allowed_avatar_mimetypes
+                and content_type not in self._profile_handler.allowed_avatar_mimetypes
+            ):
+                raise Exception("mime type not allowed for sso avatar")
+
+            # download picture
+            response = await self._http_client.request("GET", picture_https_url)
+            if response.code != 200:
+                raise Exception(
+                    "error sending GET request to provided sso avatar image"
+                )
+            image = await make_deferred_yieldable(readBody(response))
+
+            # store it in media repository
+            avatar_mxc_url = await self.media_repo.create_content(
+                str(content_type),
+                None,
+                io.BytesIO(image),  # convert image into BytesIO
+                content_length,
+                uid,
+            )
+
+            # save it as user avatar
+            await self._profile_handler.set_avatar_url(
+                uid,
+                create_requester(uid),
+                str(avatar_mxc_url),
+            )
+
+            logger.info("successfully saved the user avatar")
+        except Exception:
+            logger.exception("failed to save the user avatar")
```

Comment:
Ideally, we should only use logger.exception to report exceptions that were our (Synapse's) fault. For example: a failure in _profile_handler.set_avatar_url. 

But some of the exceptions raised above are because the remote end is at fault---those would ideally be logged as warnings. For example, a non-200 return code from the remote server.

Reply:
Understood! Changed it to `logger.warning()`